src/Avalonia.Input/TextInput/TransformTrackingHelper.cs

```diff
@@ -0,0 +1,109 @@
+        private readonly List<Visual> _propertyChangedSubscriptions = new List<Visual>();
+        
+        public TransformTrackingHelper()
+        {
+            _propertyChangedHandler = PropertyChangedHandler;
+        }
+
+        public void SetVisual(IVisual? visual)
+        {
+            Dispose();
+            _visual = visual;
+            if (visual != null)
+            {
+                visual.AttachedToVisualTree += OnAttachedToVisualTree;
+                visual.DetachedFromVisualTree -= OnDetachedFromVisualTree;
+                if (visual.IsAttachedToVisualTree)
+                    SubscribeToParents();
+                UpdateMatrix();
+            }
+        }
+        
+        public Matrix? Matrix { get; private set; }
+        public event Action? MatrixChanged;
+        
+        public void Dispose()
+        {
+            if(_visual == null)
+                return;
+            UnsubscribeFromParents();
+            _visual.AttachedToVisualTree -= OnAttachedToVisualTree;
+            _visual.DetachedFromVisualTree -= OnDetachedFromVisualTree;
+            _visual = null;
+        }
+
+        private void SubscribeToParents()
+        {
+            var visual = _visual;
+            // ReSharper disable once ConditionIsAlwaysTrueOrFalse
+            // false positive
+            while (visual != null)
+            {
+                if (visual is Visual v)
+                {
+                    v.PropertyChanged += _propertyChangedHandler;
+                    _propertyChangedSubscriptions.Add(v);
+                }
+
+                visual = visual.VisualParent;
+            }
+        }
+
+        private void UnsubscribeFromParents()
+        {
+            foreach (var v in _propertyChangedSubscriptions)
+                v.PropertyChanged -= _propertyChangedHandler;
+            _propertyChangedSubscriptions.Clear();
+        }
+
+        void UpdateMatrix()
+        {
+            Matrix? matrix = null;
+            if (_visual != null && _visual.VisualRoot != null)
+                matrix = _visual.TransformToVisual(_visual.VisualRoot);
+            if (Matrix != matrix)
+            {
+                Matrix = matrix;
+                MatrixChanged?.Invoke();
+            }
+        }
+
+        private void OnAttachedToVisualTree(object sender, VisualTreeAttachmentEventArgs visualTreeAttachmentEventArgs)
+        {
+            SubscribeToParents();
+            UpdateMatrix();
+        }
+
+        private void EnqueueForUpdate()
+        {
+            if(_queuedForUpdate)
+                return;
+            _queuedForUpdate = true;
+            Dispatcher.UIThread.Post(UpdateMatrix, DispatcherPriority.Render);
+        }
+
+        private void PropertyChangedHandler(object sender, AvaloniaPropertyChangedEventArgs e)
+        {
+            if (e.IsEffectiveValueChange && e.Property == Visual.BoundsProperty)
```

Comment:
You don't actually need to check `IsEffectiveValueChange` here - all changes that get as far as `PropertyChanged` will be effective value changes.